### ISSUE TITLE
Edge case of tag and length display

### ISSUE THREAD

**Issue by PythonTryHard** [clangen/musikcube#358]
- Version affected: [0.92.1](https://github.com/clangen/musikcube/releases/tag/0.92.1).
- OS tested against: Windows 10 x64 running the `win64` build (`winver` returned version `1903` build number `18362.900`) and Ubuntu 20.04 LTS running the `focal_amd64` build (kernel `5.4.0-37-generic`)

- Steps to replicate:
    - Download the [affected samples (2 tracks)](https://github.com/clangen/musikcube/files/4840066/Affected.Samples.zip) and load them into musikcube
    - Reload library and find the affect tracks (will be listed under `[unknown album]`)

- Expected behaviour: Tags and length displayed correctly.
- Actual behaviour: No tags were displayed, length displayed as `0:00`. musikcube defaulted to displaying the full path and filename. 

- Logs: None

- Screenshots
    - Windows: ![image](https://user-images.githubusercontent.com/31789326/85916694-7ca78e80-b87d-11ea-8180-2c6b90435b16.png)
    - Ubuntu: ![image](https://user-images.githubusercontent.com/31789326/85916719-a6f94c00-b87d-11ea-940c-70072d249dc3.png)

**Comment by PythonTryHard** [comment]
Attempted to build 97d5124, bug present as well.

**Comment by clangen** [comment]
After some debugging this appears to be a bug in `taglib` that has since been fixed upstream, but not released. We seem to be able to work around this issue by simply always compiling from source, instead of relying on the distro's supplied version.

I'm testing the changes in this branch across different platforms and architectures to make sure it doesn't cause any other problems: https://github.com/clangen/musikcube/pull/359

**Comment by PythonTryHard** [comment]
Attempted to build c1ea922 from `master` on the same Ubuntu 20.04 LTS. The problem persists, with extra cherry on top of the bug cone: It crashes, with `SIGABRT (6)` as the exit code (as shown by `zsh`) with the message being `[1] {PID} abort (core dumped)  ./musikcube`. I tried searching around for a core dump but none were found.

Will get around to building on Windows later if Visual Studio is willing to co-operate

**Comment by clangen** [comment]
Well that super disappointing -- it seems to be working properly on all operating systems on my end, but the change should be easy enough to revert. I get the feeling you may need to run the `clean-nix.sh` script in the source root and rebuild -- would you mind trying that if you get a chance?

**Comment by clangen** [comment]
Ah, I think I found something wonky in the CMake setup that's causing this issue. Let me see if I can figure out how to fix it...

**Comment by clangen** [comment]
OK, if you get a chance pull from `master`, clean, and try again. There was an issues with the CMake configuration when taglib was already installed on the OS.

**Comment by PythonTryHard** [comment]
I just deleted the repo and `git clone --recursive` a fresh one for extra precaution. Building now on my day-to-day 18.04 LTS, will build on the 20.04 LTS later (have to build it with `make -j 1` to prevent surprise thermal shutdown because it's an old one)

**Comment by PythonTryHard** [comment]
With a fresh repo, I attempted to build 3c2a06c with a fresh `git clone --recursive` and ran `sudo apt remove libtag1-dev` to be extra sure on the 20.04 LTS and it crashes when attempting to play the affected files (which is what happened and that I should have specified when I wrote [this reply](https://github.com/clangen/musikcube/issues/358#issuecomment-651841676), and the crash **when attempting to play the affected file** happened in both c1ea922 (before the change in CMake setup) and 3c2a06c (after the change in CMake setup) . I hate myself for my poor wording, and probably wasting your time debugging the CMake setup)

Edit: So yea. I'm facepalming, we're facepalming. So sorry

**Comment by clangen** [comment]
Oh no worries! There was definitely a CMake issue either way that needed to be fixed. So after pulling master and rebuilding does the metadata at least read correctly, even though playback crashes?

Playback seems fine on Windows, I'll try it on Ubuntu 20.04 this evening to see what's going on there.

**Comment by PythonTryHard** [comment]
Metadata still doesn't parse properly, sadly

**Comment by clangen** [comment]
So here's what i'm seeing on my 18.04 LTS:

![image](https://user-images.githubusercontent.com/2031666/86300765-c8797100-bbb8-11ea-83f3-b111ee531ab1.png)

I did, however, experience a couple small issues on 18.04 that didn't seem present on other distributions related to metadata indexing, so I went ahead and patched them up during lunch today.

Can you pull, and rebuild and send me the output of the `cmake .` command? It should contain the following:

```
clangen@clangen-VirtualBox:~/src/musikcube$ cmake .
...
-- [taglibmetadatareader] defining FFMPEG_ENABLED
-- [taglibmetadatareader] using bundled taglib = true
-- [taglibmetadatareader] using /home/clangen/src/musikcube/src/plugins/taglib_plugin/taglib-1.11/stage/lib/libtag.a
-- Boost version: 1.65.1
...
-- Build files have been written to: /home/clangen/src/musikcube
```

And the output of `ldd ./bin/plugins/libtaglibreader.so` -- it should look like this:

```
clangen@clangen-VirtualBox:~/src/musikcube$ ldd ./bin/plugins/libtaglibreader.so
        linux-vdso.so.1 (0x00007ffedd581000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f304181a000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f30415fd000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f3041274000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f3040ed6000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f3040cbe000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f30408cd000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f3041e3e000)
```

**Comment by clangen** [comment]
Oh, I just thought of something -- the problem may be that you need to force a meatadata refresh. You can do that by pressing `CTRL+R` anywhere in the library and selecting `rebuild library`.

**Comment by PythonTryHard** [comment]
.....ah sh/t I forgot about the refresh

**Comment by clangen** [comment]
So after a rebuild does the metadata parse? :)

**Comment by PythonTryHard** [comment]
Yes it does. Dammit me.
![image](https://user-images.githubusercontent.com/31789326/86325278-fe2a6280-bc69-11ea-8708-683a84198ac5.png)

**Comment by PythonTryHard** [comment]
As for the two commands, on the 20.04 LTS, `cmake .` returned 
```-- [build] link libraries are: Boost::system;Boost::filesystem;Boost::thread;curl;pthread;crypto;dl
-- [taglibmetadatareader] defining FFMPEG_ENABLED
-- [taglibmetadatareader] using bundled taglib = true
-- [taglibmetadatareader] using /home/r3m1/Music/musikcube/src/plugins/taglib_plugin/taglib-1.11/stage/lib/libtag.a
-- [stockencoders] defining FFMPEG_ENABLED
-- [libopenmpt] enabled = true; found library at /usr/lib/x86_64-linux-gnu/libopenmpt.so
-- [ffmpeg] enabled = true
-- [mpris] systemd found at /usr/lib/x86_64-linux-gnu/libsystemd.so
-- [mpris] setting ENABLE_MPRIS=true
-- Configuring done
-- Generating done
-- Build files have been written to: /home/r3m1/Music/musikcube
```
and `ldd ./bin/plugins/libtaglibreader.so` returned 
```
        linux-vdso.so.1 (0x00007ffec5cb1000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f0235a6b000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0235a48000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f0235867000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0235718000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f02356fd000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f023550b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0235cc5000)
```

**Comment by clangen** [comment]
Ooh, awesome, progress! Great. Does it still crash on playback though? 😬

**Comment by PythonTryHard** [comment]
Sadly it still does 😬

**Comment by clangen** [comment]
A couple follow-up questions:

1. Does the app crash or freeze? If it crashes, does it dump any message to the terminal?
2. What output driver are you using? `PulseAudio`? Does it work if you use `ALSA` instead? Does it not crash if you try `Null`?
3. Is there anything in your `~/.config/musikcube/stderr.txt` file?

**Comment by PythonTryHard** [comment]
Checked. On the 20.04 LTS I don't have core dump enabled. This is built from commit 3c2a06c

1. It crashes when attempting to playback. Here's the screenshot when it crashes
![image](https://user-images.githubusercontent.com/31789326/86407876-13da6f00-bce0-11ea-8970-c2552d5c85ad.png)
2. Yep, output driver set to `PulseAudio`. `AlsaOut` also crashes with the same post-mortem screenshot (except for PID). `Null` also crashes.
3. To be sure of no interferance, I deleted the `stderr.txt` before running `watch -n 0.5 cat stderr.txt` to monitor. Right after `musikcube` crashes (with only the affected files in the library), `stderr.txt` registered the following (driver is PulseAudio):
```
AlsaOut::AlsaOut() called
PulseOut::PulseOut() called
AlsaOut: destructor
AlsaOut: joining...
AlsaOut: thread finished
AlsaOut: closing device...
AlsaOut: destroyed.
PulseOut: opening device
pulse_blocking_stream: CHECK_VALIDITY server
pulse_blocking_stream: CHECK_VALIDITY dir
pulse_blocking_stream: CHECK_VALIDITY device
pulse_blocking_stream: CHECK_VALIDITY spec
pulse_blocking_stream: CHECK_VALIDITY channel map
pulse_blocking_stream: context is ready
pulse_blocking_stream: stream is ready
Assertion '!bus->current_slot' failed at src/libsystemd/sd-bus/sd-bus.c:2999, function bus_process_internal(). Aborting.
```

**Comment by clangen** [comment]
Whoa. Yeah, I saw this **once** on one of my Raspberry Pi devices but haven't been able to figure out how to reproduce or fix it. It's some strange problem with `mpris` / `systemd`.

If you do the following does everything work again?

1. `cd ~/src/musikcube` (or whatever)
2. `rm ./bin/plugins/libmpris.so`
3. `./bin/musikcube`

Do older versions of musikcube work fine?

**Comment by PythonTryHard** [comment]
Yep, everything worked after deleting `libmpris.so`. I'll will test the older versions later, I've been up for longer than I should have.

**Comment by clangen** [comment]
OK, I reverted a change I accidentally checked into the MPRIS plugin -- I think it will fix the issue you're having. Yikes, what a bug.

**Comment by PythonTryHard** [comment]
I tested `0.90.0`, `0.92.0` on the 20.04 machine, all three works normally without crashing on playback of the affected files. Further testing down to `0.70.0` on my day-to-day 18.04 and none had issues except for metadata parsing. 

I also attempted a fresh build on c2ed479 on both 18.04 and 20.04, both parse metadata properly and don't crash on playback of the affected tracks. While I think this is enough for this issue to be closed since it was originally about metadata, given the transpiration of events.... I'm gonna leave the call to close to you. What a bug indeed.

**Comment by clangen** [comment]
Going to close this one out, as I believe I also figured out the `sd -bus` issue I was observing on my end -- after reading some additional docs, I discovered we had a threading issue. Everything seems stable now.

**Comment by PythonTryHard** [comment]
So er...about the "fixed for next release" tag....are you gonna add it?

**Comment by clangen** [comment]
Yep, it'll be released in 0.93, which I'm working on right now. Looks like github autoclosed this issue when I merged the changes back to the master branch.

**Comment by clangen** [comment]
OK, 0.93 has been released. Closing!
